Michael

bgraham89 opened the pull request:
I've made this as a pull request so I can add comments to it

michaelsammueller committed:
Added class structure

Added the initial class structure based on the UML Class Diagram. Added class methods. Added initial functionality.

Code currently doesn't work yet, still figuring out the structure and logic.

michaelsammueller committed:
Updated class.py

Can now calculate total project cost, total design and manufacture cost, and total staff cost.

michaelsammueller committed:
Added COCOCMO

Implemented COCOMO estimation method as part of the cost estimator.

bgraham89 commented on classes.py:
```diff
@@ -0,0 +1,186 @@
+        self.title = title  # Job Title
+        self.cost_per_day = cost_per_day
+        self.employment_type = employment_type  # In-House or Agency
+        self.weeks_assigned = weeks_assigned  # Amount of weeks assigned to the job
```
We could get the estimator to automatically assign tasks to members if we did the following things:

1. Assign each staff member a maximum amount of days they can work. For example, we know the project length will be ~12 months or 52 weeks and each and most staff will work 5 days a week, so those staff could have a `workday_cap` of 52 x 5. The advantage of that would be that our programme can do the allocation of tasks for us by selecting staff for tasks from a pool of staff as long as those staff havent met their workday_cap amount of days. 
2. We could record staff member capabilities as strings in a set. So our StaffMember, constructor might look like this instead:   `def __init__(self, title, cost_per_day, employment_type, workday_cap, capabilities): `. The application could use this information as a constraint when assigning tasks.
3. Update the total_design_cost method of the ProjectEstimator class to allocate each week to a suitable member of staff, and sum the result (while tracking how many days each person has available, perhaps in a 'counter' object)

michaelsammueller replied:
That's a great idea, I'll get right to it! Thanks for having a look @bgraham89 !

michaelsammueller committed:
Update classes.py

StaffMembers are now stored as instances, not just as dictionaries with their attributes.

We now assign capabilities to each staff member, as well as required capabilities to each component.

The program assigns staff members to tasks according to their capabilities and the remaining balance of work days. The program ensures that the amount of work performed by each staff member does not exceed an overall workday_cap.

michaelsammueller committed:
Update classes.py

michaelsammueller committed:
Update classes.py

michaelsammueller commented:
I've updated the `classes.py` file according to @bgraham89 's comments.

- Every `StaffMember` instance is now stored as an object within the `ProjectEstimator` class. This allows us to access their attributes later.
- Every `StaffMember` now has a `workday_cap`, `capabilities`, and `workdays_used` attribute. This allows us to set an overall cap on their work days, their capabilities (as part of a set), and use `workdays_used` as a counter to check how many days that staff member has already worked.
- Every `HardwareComponent` and `SoftwareComponent` now has a `required_capabilities` attribute (stored as a set), which we can use to define the required capabilities for said component (f.e. "Hardware Design", "Software Design", "Hardware Manufacturing", or whatever else we want to define).
- The `total_design_cost()` and `total_manufacturing_cost()` methods iterate over each component to check their required capabilities. They then iterate over every staff member in the `project_staff` dictionary and check their capabilities. If a staff member has the required capability for the current component, then the staff member will be assigned to that component - as long as the required time does not exceed his/her work days.

Please do let me know if you see anything obvious that I have missed.

michaelsammueller committed:
Update classes.py

Fixed inconsistency. Now "days" are used as the metric to measure cost and time, as previously both days and weeks were used.

michaelsammueller commented:
Fixed an inconsistency where both "days" and "weeks" were used to measure time. Now only "days" are used to keep it consistent with cost calculation.

michaelsammueller committed:
Update classes.py

Fixed calculation issue.

michaelsammueller committed:
Updated Algorithms

Updated the cost calculation algorithms. The estimation algorithm is still pending.

michaelsammueller committed:
Update calculation.py

Added missing calculations.

michaelsammueller committed:
Update calculation.py

michaelsammueller committed:
Update calculation.py

michaelsammueller committed:
Final Adjustments & Unit Tests

Made final adjustments to the COCOMO estimation. Added comments wherever necessary to explain code. Created 'test_calculation.py' and wrote unit tests for the code.

michaelsammueller committed:
Remove old files

Removed old files.

michaelsammueller committed:
Update calculation.py

michaelsammueller committed:
Latest push

michaelsammueller committed:
fixed merge conflicts
